ethcore/sync/src/block_sync.rs

```diff
@@ -439,22 +452,28 @@ impl BlockDownloader {
 					trace_sync!(self, "Searching common header from the last round {} ({})", self.last_imported_block, self.last_imported_hash);
 				} else {
 					let best = io.chain().chain_info().best_block_number;
+					let best_hash = io.chain().chain_info().best_block_hash;
 					let oldest_reorg = io.chain().pruning_info().earliest_state;
 					if self.block_set == BlockSet::NewBlocks && best > start && start < oldest_reorg {
 						debug_sync!(self, "Could not revert to previous ancient block, last: {} ({})", start, start_hash);
-						self.reset();
+						self.reset_to_block(&best_hash, best);
 					} else {
 						let n = start - cmp::min(self.retract_step, start);
-						self.retract_step *= 2;
-						match io.chain().block_hash(BlockId::Number(n)) {
-							Some(h) => {
-								self.last_imported_block = n;
-								self.last_imported_hash = h;
-								trace_sync!(self, "Searching common header in the blockchain {} ({})", start, self.last_imported_hash);
-							}
-							None => {
-								debug_sync!(self, "Could not revert to previous block, last: {} ({})", start, self.last_imported_hash);
-								self.reset();
+						if n == 0 {
+							debug_sync!(self, "Header not found, bottom line reached, resetting, last imported: {}", self.last_imported_hash);
+							self.reset_to_block(&best_hash, best);
```

Comment:
Unfortunately, here the calling of simple reset is not enough, because it resets counters (last_imported_block) to the current round. In this case the rounds went too far from and simple rollback inside round doesn't help. reset_to_block here actually resets block sync to its initial state